lib/active_model/serializer/adapter/json.rb

```diff
@@ -4,44 +4,130 @@ class ActiveModel::Serializer::Adapter::Json < ActiveModel::Serializer::Adapter
 
         def serializable_hash(options = nil)
           options ||= {}
-          if serializer.respond_to?(:each)
-            result = serializer.map { |s| FlattenJson.new(s).serializable_hash(options) }
-          else
-            hash = {}
 
-            core = cache_check(serializer) do
-              serializer.attributes(options)
+          hash =
+            if serializer.respond_to?(:each)
+              serializable_hash_for_collection(options)
+            else
+              serializable_hash_for_single_resource
             end
 
-            serializer.associations.each do |association|
-              serializer = association.serializer
-              opts = association.options
-
-              if serializer.respond_to?(:each)
-                array_serializer = serializer
-                hash[association.key] = array_serializer.map do |item|
-                  cache_check(item) do
-                    item.attributes(opts)
-                  end
-                end
-              else
-                hash[association.key] =
-                  if serializer && serializer.object
-                    cache_check(serializer) do
-                      serializer.attributes(options)
-                    end
-                  elsif opts[:virtual_value]
-                    opts[:virtual_value]
-                  end
-              end
-            end
-            result = core.merge hash
-          end
+          { root => hash }
+        end
+
+        def serializable_hash_for_collection(options = nil)
+          serializer.map { |s| FlattenJson.new(s).serializable_hash(options) }
```

Comment:
I think we have way too many PRs changing the same stuff right now. Haha. We just need to get them merged, and handle whatever conflicts arise.